File: src/main/java/com/clearcont/clearcontapp/views/routes/FornecedoresView.java
package com.clearcont.clearcontapp.views.routes;

import com.clearcont.clearcontapp.helpers.CookieFactory;
import com.clearcont.clearcontapp.helpers.MonthAndCompany;
import com.clearcont.clearcontapp.model.Balancete;
import com.clearcont.clearcontapp.model.Empresa;
import com.clearcont.clearcontapp.model.Responsavel;
import com.clearcont.clearcontapp.repository.CustomerContabilRepository;
import com.clearcont.clearcontapp.repository.EmpresaRepository;
import com.clearcont.clearcontapp.repository.ResponsavelRepository;
import com.clearcont.clearcontapp.service.BalanceteService;
import com.clearcont.clearcontapp.service.CustomerContabilService;
import com.clearcont.clearcontapp.views.components.GridCustomer;
import com.clearcont.clearcontapp.views.components.GridFornecedores;
import com.clearcont.clearcontapp.views.main.MainLayout;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.html.H1;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.notification.Notification;
import com.vaadin.flow.component.orderedlayout.FlexLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.router.PageTitle;
import com.vaadin.flow.router.Route;
import com.vaadin.flow.server.VaadinService;
import jakarta.annotation.security.RolesAllowed;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Route(value = "fornecedores", layout = MainLayout.class)
@PageTitle("Fornecedores")
@RolesAllowed("ADMIN")
@Getter
@Setter
@Slf4j
public class FornecedoresView extends FlexLayout implements MonthAndCompany {
    private String month;
    private Empresa empresa;

    public FornecedoresView(CustomerContabilService customerContabilRepository, EmpresaRepository empresaRepository, BalanceteService balanceteService, ResponsavelRepository responsavelRepository) {
        initializeView(customerContabilRepository, empresaRepository, balanceteService, responsavelRepository);
    }

    private void initializeView(CustomerContabilService customerContabilRepository, EmpresaRepository empresaRepository, BalanceteService balanceteService, ResponsavelRepository responsavelRepository) {
        getCompany(empresaRepository, empresa -> getMonth(month -> {
                    if (empresa == null) {
                        Notification.show("Selecione uma empresa e periodo");
                        UI.getCurrent().navigate("/");
                    }

                    CookieFactory cookieFactory = new CookieFactory(VaadinService.getCurrentResponse());
                    Long responsavelID = cookieFactory.getCookieInteger("responsavel-id");
                    Responsavel responsavel = responsavelRepository.findById(responsavelID).orElseThrow();
                    Long empresaId = empresa.getId();
                    List<Balancete> balanceteData = balanceteService.getByCompanyAndPeriod(empresaId, month, LocalDate.now().getYear());
                    log.info("Empresa selecionada: " + empresa.getNomeEmpresa());
                    GridFornecedores gridCustomer = new GridFornecedores(customerContabilRepository, balanceteData, responsavel);
                    H1 clientes = new H1("Fornecedores");
                    Span span = new Span(empresa.getNomeEmpresa());
                    Span subtitle = new Span("Selecione um balancete do periodo: " + month + " " + LocalDate.now().getYear());
                    VerticalLayout verticalLayout = new VerticalLayout(
                            new VerticalLayout(clientes, span, subtitle),
                            gridCustomer
                    );
                    add(verticalLayout);
                }
        ));
    }
}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/views/routes/FornecedoresView.kt
package br.com.clearcont.clearcontwebapp.views.routes

import br.com.clearcont.clearcontwebapp.utils.helpers.CookieFactory
import br.com.clearcont.clearcontwebapp.utils.helpers.MonthAndCompany
import br.com.clearcont.clearcontwebapp.utils.helpers.Period.getMonthByPeriodoString
import br.com.clearcont.clearcontwebapp.utils.helpers.createTitle
import br.com.clearcont.clearcontwebapp.models.Empresa
import br.com.clearcont.clearcontwebapp.models.enums.TipoConta.FORNECEDOR
import br.com.clearcont.clearcontwebapp.repositories.EmpresaRepository
import br.com.clearcont.clearcontwebapp.repositories.ResponsavelRepository
import br.com.clearcont.clearcontwebapp.services.impl.BalanceteService
import br.com.clearcont.clearcontwebapp.services.impl.ComposicaoLancamentosContabeisService
import br.com.clearcont.clearcontwebapp.utils.helpers.verifySelectedCompanyAndMonthExistAndNavigate
import br.com.clearcont.clearcontwebapp.utils.shared.RESPONSAVEL_ID
import br.com.clearcont.clearcontwebapp.views.components.GridFornecedores
import br.com.clearcont.clearcontwebapp.views.components.MainLayout
import com.vaadin.flow.component.UI
import com.vaadin.flow.component.html.Span
import com.vaadin.flow.component.notification.Notification
import com.vaadin.flow.component.notification.NotificationVariant
import com.vaadin.flow.component.orderedlayout.FlexLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.router.PageTitle
import com.vaadin.flow.router.Route
import com.vaadin.flow.server.VaadinResponse
import jakarta.annotation.security.RolesAllowed
import java.time.LocalDate
import java.util.logging.Logger

@Route(value = "fornecedores", layout = MainLayout::class)
@PageTitle("Fornecedores")
@RolesAllowed("ADMIN")
class FornecedoresView(
    composicaoService: ComposicaoLancamentosContabeisService,
    empresaRepository: EmpresaRepository,
    balanceteService: BalanceteService,
    responsavelRepository: ResponsavelRepository
) : FlexLayout(), MonthAndCompany {
    override var month: String? = null
    override lateinit var empresa: Empresa
    var log: Logger = Logger.getLogger(javaClass.name)


    init {
        initializeView(composicaoService, empresaRepository, balanceteService, responsavelRepository)
    }

    private fun initializeView(
        customerContabilRepository: ComposicaoLancamentosContabeisService,
        empresaRepository: EmpresaRepository,
        balanceteService: BalanceteService,
        responsavelRepository: ResponsavelRepository
    ) {
        this.getCompany(empresaRepository) { empresa: Empresa? ->
            getMonth { month: String? ->
                verifySelectedCompanyAndMonthExistAndNavigate(empresa, month)
                val cookieFactory = CookieFactory(VaadinResponse.getCurrent())
                val responsavelID = cookieFactory.getCookieInteger(RESPONSAVEL_ID)
                val responsavel = responsavelRepository.findById(responsavelID).orElseThrow()
                val empresaId = empresa!!.id
                val balanceteData =
                    balanceteService.filterClassification(empresaId!!, month!!, LocalDate.now().year, FORNECEDOR)
                log.info("Empresa selecionada: ${empresa.nomeEmpresa}")
                log.info("Tamanho balancete: $balanceteData")

                val gridCustomer = GridFornecedores(
                    customerContabilRepository,
                    balanceteData,
                    responsavel,
                    getMonthByPeriodoString(month),
                    balanceteService,
                )

                val clientes = createTitle("Fornecedores").apply { width = "50%" }
                val span = Span(empresa.nomeEmpresa)
                val subtitle = Span("Selecione uma conta do periodo: " + month + " " + LocalDate.now().year)
                val titleLayout = VerticalLayout(clientes, span, subtitle)
                val verticalLayout = VerticalLayout(titleLayout, gridCustomer)

                add(verticalLayout)
            }
        }
    }

}
